dagshub/models/model_locator.py

```diff
@@ -0,0 +1,320 @@
+
+
+class ModelNotFoundError(Exception):
+    def __str__(self):
+        return "Could not find model"
+
+
+@dataclass
+class ModelYamlStruct:
+    model_dir: str
+
+
+class StorageType(IntEnum):
+    Repo = auto()
+    DagshubStorage = auto()
+    Bucket = auto()
+
+
+class ModelLocator:
+    def __init__(
+        self,
+        repo: Optional[Union[str, RepoAPI]] = None,
+        path: Optional[str] = None,
+        bucket: Optional[str] = None,
+        git_ref: Optional[str] = None,
+        mlflow_model: Optional[str] = None,
+        mlflow_artifact: Optional[str] = None,
+        download_dest: Optional[Union[str, os.PathLike]] = None,
+        download_type: Literal["lazy", "eager"] = "eager",
+    ):
+        self._repo = repo
+        self.path = path
+        self.bucket = bucket
+        self._git_ref = git_ref
+        self.mlflow_model = mlflow_model
+        self.mlflow_artifact = mlflow_artifact
+        self._download_dest = download_dest
+        self.download_type = download_type
+
+    @cached_property
+    def repo_api(self) -> RepoAPI:
+        if isinstance(self._repo, RepoAPI):
+            return self._repo
+        if self._repo is None:
+            repo_api, git_ref = determine_repo()
+            if self._git_ref is None:
+                self._git_ref = git_ref
+            return repo_api
+        return RepoAPI(self._repo)
+
+    @property
+    def git_ref(self):
+        if self._git_ref is None:
+            self._git_ref = self.repo_api.default_branch
+        return self._git_ref
+
+    @cached_property
+    def download_destination(self) -> Path:
+        if self._download_dest is not None:
+            return Path(self._download_dest)
+        return Path(sanitize_filepath(os.path.join(Path.home(), "dagshub", "models", self.repo_api.full_name)))
+
+    @cached_property
+    def repo_storages(self) -> List[StorageAPIEntry]:
+        return self.repo_api.get_connected_storages()
+
+    def _handle_path(self, path: str) -> Tuple[str, StorageType]:
+        """
+        Handles user-printed path.
+        Returns a resulting path that can be queried from DagsHub + boolean for whether it's a repo or a bucket path
+        If path is a repo path, returns (path as is, StorageType.Repo)
+        If path is a storage path, ({actual path in storage} + StorageType.Bucket/DagshubStorage)
+        Storage path resolution is based on bucket name, if the path starts with a name of an existing bucket,
+        assume that it's a storage path
+        """
+        str_path = str(path)
+
+        if "/" not in str_path:
+            return path, StorageType.Repo
+
+        if str_path.startswith("dagshub_storage/"):
+            return str_path[len("dagshub_storage/"):], StorageType.DagshubStorage
+        for storage in self.repo_storages:
+            if str_path.startswith(f"{storage.name}/"):
+                bucketPath = f"{storage.protocol}/{str_path}"
+                return bucketPath, StorageType.Bucket
```

Comment:
User here needs some advanced knowledge of necessary naming convention.

Reply:
Yea, that's true, and right now it's not consistent with other ways we use storages.
This is the format that we'll be using later when we reimplement install_hooks pathing, so the hope is that it will be consistent at that point and they wouldn't need to look anything up.

The dir tree after that change should look like this basically:
```
.
|-- README.md
|-- dagshub_storage
|-- my_integrated_bucket
`-- src
```

and the path for the model should be intuitive